### ISSUE TITLE
Unstable retrieval of energy prices

### ISSUE THREAD

**Issue by gskjold** [gskjold/AmsToMqttBridge#198]
- Sometimes does not refresh data
- Sometimes uses data from wrong day after midnight

**Comment by ArnieO** [comment]
It is probably already taken into account, but a potential source of issues is non-response or wrong response from the ENTSO-E API. I have seen that it is not quite stable.

**Comment by gskjold** [comment]
Yes, I have seen that as well. I have a stand alone code to retrieve prices for my home automation, so I can run a comparison with it for a while to see if it is API or our code.

**Comment by tronde-ams** [comment]
Is "verbose" the preferred logging to catch errors, or should I choose something else? I will try to log until you close this topic.

Regarding the API. 
I understand you only retrieve prices once a day. Is it a problem to retrieve them every hour? The limit for the API seems to be 400 polls per minute.

**Comment by ArnieO** [comment]
I have seen periods (not often) where the API failed systematically over several days. The code should maybe have a way to signal «failed to retreive prices».

**Comment by gskjold** [comment]
> Is "verbose" the preferred logging to catch errors, or should I choose something else? I will try to log until you close this topic.

Only INFO and DEBUG level messages logged from ENTSO-E code, so DEBUG is enough.



> I understand you only retrieve prices once a day. Is it a problem to retrieve them every hour? The limit for the API seems to be 400 polls per minute.

It is possible to get prices every hour, but I see no reason to do so. A lot of resources are used to establish SSL connection with remote server, which could impact other (more important) functions of the software. In fact,I've seen several times that the device reboots when establishing connection with the server.

**Comment by tronde-ams** [comment]
This is what is logged as "debug".
It is the same as yesterday when the prices re-apperared at midnight. Does not seem to give any useful infrmation about changes? I deleted the old log, but it was the same information shown.

<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(D) (loop)(C1) Used 0 ms to update temperature
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(D) (loop)(C1) Used 0 ms to update temperature
<ESC>[0m(D) (AmsWebServer)Serving /data.json over http...
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18
<ESC>[0m(I) (readHanPort)(C1) Valid data, start at byte 18

**Comment by gskjold** [comment]
There will only be logging about prices when they are fetched, right after boot and at 13.00. Also some logging at midnight when prices are rotated.

**Comment by tronde-ams** [comment]
Ok, Is it possible to see from the verbose log that (or why) prices disappear? I have no personal interest in logging. Only to assist in finding the bug.

**Comment by gskjold** [comment]
Setting to debug and catching output from 13.00 to 13.15 and from 23.59 to 00.01 is the only way to see if it is doing what it should. I am not very hopeful though, this bug could be a logic issue that we could not catch by watching logs.

**Comment by ArnieO** [comment]
Could it be an idea to avoid reading prices shortly after 13:00?

- Yes, prices should be published at 13:00 but I have seen days with delays, so in an automation I implemented some years back I decided to read ENTSO-E prices later in the evening.
- Traffic on the ENTSO-E server is probably higher shortly after prices are published. Could be an argument for waiting. 
- Maybe even consider randomizing the time of reading to avoid that (in a not too distant future! 😉 ) thousands of AmsToMqttBridge devices poll the server at exactly the same time.
